lib/AST/NameLookup.cpp

```diff
@@ -1195,12 +1199,42 @@ TinyPtrVector<ValueDecl *> NominalTypeDecl::lookupDirect(
                                                   DeclName name,
                                                   bool ignoreNewExtensions) {
   RecursiveSharedTimer::Guard guard;
-  if (auto s = getASTContext().Stats) {
+  ASTContext &ctx = getASTContext();
+  if (auto s = ctx.Stats) {
     ++s->getFrontendCounters().NominalTypeLookupDirectCount;
     guard = s->getFrontendRecursiveSharedTimers()
                 .NominalTypeDecl__lookupDirect.getGuard();
   }
 
+  DEBUG(llvm::dbgs() << getNameStr() << ".lookupDirect(" << name << ")"
+        << ", lookupTable.getInt()=" << LookupTable.getInt()
+        << ", hasLazyMembers()=" << hasLazyMembers()
+        << "\n");
+
+  if (ctx.LangOpts.NamedLazyMemberLoading &&
+      !LookupTable.getInt() &&
+      hasLazyMembers()) {
+    // The lookup table (containing all loaded members) has not yet been built;
+    // we *might* be able to avoid loading all members, just focus on the ones
+    // matching the name we were asked for.
+    TinyPtrVector<ValueDecl *> results;
+    auto contextInfo =
+        ctx.getOrCreateLazyIterableContextData(this,
+                                               /*lazyLoader=*/nullptr);
+    if (contextInfo->loader->loadNamedMembers(this, name,
```

Comment:
Hmmm, this won't be sufficient, because `lookupDirect` is supposed to return all of the members of the given name across the nominal type and all of its extensions, but the `contextInfo` is only associated with the nominal type declaration itself. At worst, we will have to poll several different sources to gather all of the declarations that match a given name:

* The Clang importer, which can answer this question by looking into all of the Swift name lookup tables,
* The Swift module loader, which can answer this question once we've added some name lookup tables for nominal types, and
* Walking the members of any extensions of the type that we parsed into the current module

We have to combine the results from all of these sources to get the right results from name lookup.

Reply:
Yes, those are reserved for phases 2/N, 3/N and so forth in this work series. This version literally just does ObjC `@protocol` members (and I mistakenly thought imported `@protocol`s were not themselves extendable on the swift side). I'll add a check for "not extended by swift" for this variant; will return to this when we handle the swift side (i.e. have extended swiftmodules with member lookup tables too).